Add p2pcam integration

indykoning opened the pull request:
## Description:

Add support for generic cheap chinese p2pcamers operating locally under a UDP protocol.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10767

## Example entry for `configuration.yaml` (if applicable):
```yaml
camera:
  - platform: p2pcam
    ip_address: 192.168.178.9
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

indykoning committed:
Added p2pcam code

indykoning committed:
Ran black formatter

indykoning committed:
Added periods to docblocks.

indykoning committed:
Add p2pcam to coveragerc

fabaff commented on homeassistant/components/p2pcam/camera.py:
```diff
@@ -0,0 +1,75 @@
+        return self._name
+
+
+def get_host_ip():
```
Besides that the code is very hard to read I doubt that you will get the right IP address if there are multiple interfaces present and connected. 

At first glance I would says get the IP address from the `discovery` integration which is triggered by the `zeroconf` integration (part of `default_config`).

fabaff commented:
Perhaps it worth to handle `p2pcam` like a component.

indykoning committed:
Use home assistant utils to get ip

indykoning commented:
I tried to get it set up as a integration however it does not pass through validation:
```
CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.All(
            cv.ensure_list,
            [
                vol.Schema(
                    {
                        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
                        vol.Optional(CONF_HOST): cv.string,
                        vol.Required(CONF_IP_ADDRESS): cv.string,
                    }
                )
            ],
        )
    },
    extra=vol.ALLOW_EXTRA,
)
```
When i have this for the config validation i end up getting this message when validating:
```
Invalid config for [p2pcam]: [ip_address] is an invalid option for [p2pcam]. 
Check: p2pcam->ip_address. (See ?, line ?). 
Please check the docs at https://home-assistant.io/integrations/p2pcam/
```
This is what i'm using as my config:
```
p2pcam:
  - ip_address: IP_ADDRESS
```

indykoning committed:
Update file to rerun tests.

I noticed that after the update to `0.101.2` it causes home assistant to crash locally.
I want to test if this is just my installation or if it's the package.

indykoning committed:
Bump p2pcam version

indykoning committed:
Bump p2pcam

springstan commented on homeassistant/components/p2pcam/const.py:
```diff
@@ -0,0 +1,3 @@
+"""Constants for the p2pcam component."""
+
+DEFAULT_NAME = "p2pcam"
```
If you only have one constant move it to the camera.py file and delete this one.

springstan commented on homeassistant/components/p2pcam/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "domain": "p2pcam",
+  "name": "p2pcam",
+  "documentation": "https://www.home-assistant.io/components/p2pcam",
```
Replace components with integrations in your documentation link.

springstan commented on homeassistant/components/p2pcam/camera.py:
```diff
@@ -0,0 +1,54 @@
+"""P2PCamera integration."""
+import logging
+
+import p2pcam as p2pcam_req
```
Only import the used class, e.g.
```python
from p2pcam import P2PCam
```

springstan commented on homeassistant/components/p2pcam/camera.py:
```diff
@@ -0,0 +1,54 @@
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Register camera."""
+    async_add_entities([P2PCam(hass, config)])
```
What happens if you have multiple cameras?

springstan commented on homeassistant/components/p2pcam/camera.py:
```diff
@@ -0,0 +1,54 @@
+
+        self._name = config[CONF_NAME]
+        self._host_ip = config[CONF_HOST]
+        self._target_ip = config[CONF_IP_ADDRESS]
```
Use config.get() for optional and use config[key] for required parameters.

balloob commented on homeassistant/components/p2pcam/camera.py:
```diff
@@ -0,0 +1,54 @@
+
+    async def async_camera_image(self):
+        """Retrieve the camera image."""
+        return self.camera.retrieveImage()
```
This looks like you're doing I/O inside the event loop. Please implement `def camera_image` instead.

balloob commented:
This PR seems to have gone stale. Closing it.
